package/MDAnalysis/lib/util.py

```diff
@@ -2417,3 +2385,26 @@ def check_box(box):
     if np.all(box[3:] == 90.):
         return 'ortho', box[:3]
     return 'tri_vecs', triclinic_vectors(box)
+
+
+@deprecate(release="2.2.0", remove="3.0.0",
+           message="lib.util.which to be remove from 3.0.0 release")
+def which(program):
```

Comment:
no we don't want to re-implement a new version of which, in the future folks will just use `shutil.which`. The deprecation needs to go on the old version of `lib.util.which`.

Reply:
I moved the "which" function at EOF since the deprecate function was declared above in the code.

Reply:
@IAlibay I think we should remove the old code and replace it with something like
```python
# TODO: Remove in 3.0.0
@deprecate(release="2.2.0", remove="3.0.0",
           message="lib.util.which() will be removed in 3.0.0; use shutil.which() instead")
def which(program):
   """Determine full path of executable `program` on :envvar:`PATH`."""
   return shutil.which(program)
```

Reply:
I'm not sure I agree, in my opinion if you're doing a deprecation like this you're better off keeping the old code in case there's some very specific difference between the two `which` implementations that somehow someone relies on.

Also going with a change of the underlying method just for the sake of removing it later also just seems like unnecessary line changes. If the old one is broken why not just add the deprecation decorator to it and leave it? Are there any benefits here which I'm missing?

Reply:
My thinking was that we can remove code and as long as it still passes the tests, it provides what we promised.

You're, of course, right that there could be subtle differences. In this particular case, I doubt anyone cares for such a utility function. But sure, we can just keep the old one, your call.

Reply:
So the call I'm going to make here is that whatever we do we shouldn't just move the `which` method to the bottom of the file. The line diff here is going to be confusing if someone comes along and tries to work out the history of this file.

I have a preference for just keeping the old method for now but that's mostly for the sake of simplicity. That being said, I don't really mind if we go @orbeckst's way, I'm happy to approve either as long as we can follow the above requirement.

Reply:
Note: if it's a case of deprecate being defined within the same file, just adding the warning manually would be a preferred option, that's fewer line changes.

Reply:
Let's just keep the old code for `lib.util.which()` and _not_ replace it as I suggested in https://github.com/MDAnalysis/mdanalysis/pull/3667#discussion_r878671837.

@IAlibay has a strong opinion on not rearranging the file so `which` needs to go back to its original place and the deprecation message added manually.